### PR TITLE
Fix #3241 UX/UI - Introduce PrivateSend tab which allows to spend fully mixed coins only

### LINKED ISSUE #3241: 0.15.0.0-rc1 UI/UX Send tab move Privatesend checkbox
### Describe the issue
v15 has a much cleaner look, but the privatesend checkbox is still easily overlooked, and can result in a user sending regular funds by accident.

On the Send tab, move PrivateSend checkbox and Balance to the left side, just above the Send button and below "Note: Not enough data for fee estimation..."

When the user goes to press send, they have a better chance of seeing the privatesend box, and remembering to check it before sending. Privacy mistakes are easy to make and the consequences can be severe in some countries.. This will help reduce mistakes.

### PR DESCRIPTION
**Pull request by 10xcryptodev**
Apply the changes suggested on #3241

### CODE REVIEW

**Comment by UdjinM6** [comment]
Looks good code-wise but I don't like the new layout tbh.

**Comment by 10xcryptodev** [comment]
@UdjinM6 agree with you. Another suggestion that i thought is use the old layout, remove the PrivateSend checkbox and add the option in confirmation dialog Yes (PrivateSend). It will be cleaner and the user will always see the option before send

**Comment by UdjinM6** [comment]
Simply removing PS checkbox won't work because:
- it also switches "Balance" from "All" to "PS only" numbers, you'd have to show both balances if you remove it;
- it affects numbers in CoinControl labels (PS txes can't have change, "fee"/"after fee" are adjusted accordingly), duplicating these would be a mess.

IMO the GUI is already pretty clear + there is another label about funds you are going to use (mixed/no-mixed) in confirmation box. The only possible tweak I can think about is that we could add another button "PrivateSend" right next to "Send" which would internally call the same `on_sendButton_clicked()` but would be disabled if PS checkbox is off (and "Send" would be disabled if PS checkbox is on). Not sure if that would it make things clearer or the other way around though.

**Comment by 10xcryptodev** [comment]
@UdjinM6 PrivateSend changes the available balance and inputs, so it should be the first step the user should choose. I thought 2 possibilities:
1. Create a new PrivateSend tab between Send and Receive that will use the same UI (sendcoinsdialog.ui) and behind will act "as the checkbox is activated"
2. Ask the user when enter the Send tab if want to Send or PrivateSend using a confirmation dialog, the checkbox could stay at the same place, but the user will always be asked the option

Personally i prefer the first option, it will be clear and simple and give us more possibilities if we want to add more PrivateSend features in the future

**Comment by UdjinM6** [comment]
Hmm.... yeah, I would probably go with (1) assuming it's going to reuse the same code/ui.

**Comment by 10xcryptodev** [comment]
@UdjinM6 implemented the number 1 solution, reusing the same UI

File: src/qt/coincontroldialog.cpp [hunk 1]
```diff
@@ -440,16 +440,6 @@ void CoinControlDialog::viewItemChanged(QTreeWidgetItem* item, int column)
...
-
-    // TODO: Remove this temporary qt5 fix after Qt5.3 and Qt5.4 are no longer used.
-    //       Fixed in Qt5.5 and above: https://bugreports.qt.io/browse/QTBUG-43473
-#if QT_VERSION >= 0x050000
-    else if (column == COLUMN_CHECKBOX && item->childCount() > 0)
-    {
-        if (item->checkState(COLUMN_CHECKBOX) == Qt::PartiallyChecked && item->child(0)->checkState(COLUMN_CHECKBOX) == Qt::PartiallyChecked)
-            item->setCheckState(COLUMN_CHECKBOX, Qt::Checked);
-    }
-#endif
```
[the remark this answers is not in the document]
Comment (10xcryptodev) [reply src/qt/coincontroldialog.cpp:452]: yes, is unrelated. I will revert and open a new PR for this TODO

File: src/qt/forms/sendcoinsdialog.ui [hunk 2]
```diff
@@ -1209,23 +1209,7 @@
...
-         </property>
-         <property name="text">
-          <string>PrivateSend</string>
-         </property>
-         <property name="checked">
-          <bool>false</bool>
-         </property>
-        </widget>
-       </item>
+       </property>
```
Comment (UdjinM6) [inline src/qt/forms/sendcoinsdialog.ui:1212]: linter fails here
```suggestion
       </property>
```

**Comment by PastaPastaPasta** [comment]
Could we get some screen-shots of the new format etc?

File: src/qt/coincontroldialog.cpp [hunk 3]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
-            itemOutput->setData(COLUMN_CONFIRMATIONS, Qt::UserRole, QVariant((qlonglong)out.nDepth));
+            if ((coinControl()->IsUsingPrivateSend() && nRounds >= privateSendClient.nPrivateSendRounds) || !(coinControl()->IsUsingPrivateSend())) {
+                nSum += out.tx->tx->vout[out.i].nValue;
+                nChildren++;
+
+                CCoinControlWidgetItem* itemOutput;
+                if (treeMode)
+                    itemOutput = new CCoinControlWidgetItem(itemWalletAddress);
+                else
+                    itemOutput = new CCoinControlWidgetItem(ui->treeWidget);
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:710]: Please add brackets to be in line with the style guide

File: src/qt/coincontroldialog.cpp [hunk 4]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
+                itemOutput->setFlags(flgCheckbox);
+                itemOutput->setCheckState(COLUMN_CHECKBOX, Qt::Unchecked);
+
+                // address
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:714]: This comment seems unneeded

File: src/qt/coincontroldialog.cpp [hunk 5]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
+                CTxDestination outputAddress;
+                QString sAddress = "";
+                if (ExtractDestination(out.tx->tx->vout[out.i].scriptPubKey, outputAddress)) {
+                    sAddress = QString::fromStdString(EncodeDestination(outputAddress));
+
+                    // if listMode or change => show dash address. In tree mode, address is not shown again for direct wallet address outputs
+                    if (!treeMode || (!(sAddress == sWalletAddress)))
+                        itemOutput->setText(COLUMN_ADDRESS, sAddress);
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:722]: add brackets

File: src/qt/coincontroldialog.cpp [hunk 6]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
+
+                    itemOutput->setToolTip(COLUMN_ADDRESS, sAddress);
+                }
 
-            // transaction hash
-            uint256 txhash = out.tx->GetHash();
-            itemOutput->setText(COLUMN_TXHASH, QString::fromStdString(txhash.GetHex()));
+                // label
+                if (!(sAddress == sWalletAddress)) // change
+                {
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:729]: adjust brackets
```suggestion
                if (!(sAddress == sWalletAddress)) { // change
```

File: src/qt/coincontroldialog.cpp [hunk 7]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
+                    // tooltip from where the change comes from
+                    itemOutput->setToolTip(COLUMN_LABEL, tr("change from %1 (%2)").arg(sWalletLabel).arg(sWalletAddress));
+                    itemOutput->setText(COLUMN_LABEL, tr("(change)"));
+                } else if (!treeMode) {
+                    QString sLabel = model->getAddressTableModel()->labelForAddress(sAddress);
+                    if (sLabel.isEmpty())
+                        sLabel = tr("(no label)");
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:736]: adjust brackets

File: src/qt/coincontroldialog.cpp [hunk 8]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
+                // date
+                itemOutput->setText(COLUMN_DATE, GUIUtil::dateTimeStr(out.tx->GetTxTime()));
+                itemOutput->setToolTip(COLUMN_DATE, GUIUtil::dateTimeStr(out.tx->GetTxTime()));
+                itemOutput->setData(COLUMN_DATE, Qt::UserRole, QVariant((qlonglong)out.tx->GetTxTime()));
+
+                // PrivateSend rounds
+                if (nRounds >= 0 || LogAcceptCategory(BCLog::PRIVATESEND))
+                    itemOutput->setText(COLUMN_PRIVATESEND_ROUNDS, QString::number(nRounds));
+                else
+                    itemOutput->setText(COLUMN_PRIVATESEND_ROUNDS, tr("n/a"));
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:754]: adjust brackets

File: src/qt/coincontroldialog.cpp [hunk 9]
```diff
@@ -696,87 +696,87 @@ void CoinControlDialog::updateView()
...
-             // disable locked coins
-            if (model->isLockedCoin(txhash, out.i))
-            {
-                COutPoint outpt(txhash, out.i);
-                coinControl()->UnSelect(outpt); // just to be sure
-                itemOutput->setDisabled(true);
-                itemOutput->setIcon(COLUMN_CHECKBOX, QIcon(":/icons/lock_closed"));
+                // set checkbox
+                if (coinControl()->IsSelected(COutPoint(txhash, out.i)))
+                    itemOutput->setCheckState(COLUMN_CHECKBOX, Qt::Checked);
```
Comment (PastaPastaPasta) [inline src/qt/coincontroldialog.cpp:778]: add brackets

File: src/qt/walletframe.cpp [hunk 10]
```diff
@@ -145,6 +145,13 @@ void WalletFrame::gotoSendCoinsPage(QString addr)
         i.value()->gotoSendCoinsPage(addr);
 }
 
+void WalletFrame::gotoPrivateSendCoinsPage(QString addr)
+{
+    QMap<QString, WalletView*>::const_iterator i;
+    for (i = mapWalletViews.constBegin(); i != mapWalletViews.constEnd(); ++i)
+        i.value()->gotoPrivateSendCoinsPage(addr);
```
Comment (PastaPastaPasta) [inline src/qt/walletframe.cpp:152]: add brackets, use auto for iterator
```suggestion
    for (auto i = mapWalletViews.constBegin(); i != mapWalletViews.constEnd(); ++i) {
        i.value()->gotoPrivateSendCoinsPage(addr);
    }
```

File: src/qt/walletview.cpp [hunk 11]
```diff
@@ -237,6 +237,16 @@ void WalletView::gotoReceiveCoinsPage()
 
 void WalletView::gotoSendCoinsPage(QString addr)
 {
+    sendCoinsPage->setPrivateSend(false);
+    setCurrentWidget(sendCoinsPage);
+
+    if (!addr.isEmpty())
+        sendCoinsPage->setAddress(addr);
```
Comment (PastaPastaPasta) [inline src/qt/walletview.cpp:244]: add brackets

File: src/qt/coincontroldialog.cpp (hunk 4 above)
Comment (UdjinM6) [reply src/qt/coincontroldialog.cpp:714]: This one is from original code though

Comment (PastaPastaPasta) [reply src/qt/coincontroldialog.cpp:714]: It seems that most of the other comments are removed

Comment (UdjinM6) [reply src/qt/coincontroldialog.cpp:714]: Hah? Try reviewing with `?w=1`, they are all in place as far as I can see.

Comment (PastaPastaPasta) [reply src/qt/coincontroldialog.cpp:714]: Fair

**Comment by 10xcryptodev** [comment]
@PastaPastaPasta code formating updated

Screenshots:
Screenshot from 2020-04-26 21-34-11

New PrivateSend tab, shows only the mixed balance and mixed outputs in coin control
Screenshot from 2020-04-26 21-34-54

This way will be clear for the user that is using PrivateSend in his first action selecting the type of sending. Also before this change, the user could select PrivateSend checkbox and select a not mixed output in coin control, that could cause confusion too

**Comment by UdjinM6** [comment]
>Also before this change, the user could select PrivateSend checkbox and select a not mixed output in coin control, that could cause confusion too.

We have a warning for this case https://github.com/dashpay/dash/blob/develop/src/qt/coincontroldialog.cpp#L430-L436 (which I think can be dropped now btw since this should be no longer possible) but it's much better to hide non-mixed inputs completely, I like that.

**Comment by UdjinM6** [comment]
Hmm... Interesting. So switching from Send to PrivateSend tab with non-mixed inputs selected via CoinControl on the former results in some new confusing situation when it says N inputs selected but you can't see them cause they are not in the list. 🤔 PS would still fail as it should saying "not enough mixed funds" so it's not that bad but maybe we could also clear non-mixed inputs from selection when CoinControl is opened via PS tab?

**Comment by 10xcryptodev** [comment]
@UdjinM6 great! I will check to clear the selections when changing the tab and remove the unnecessary warning

**Comment by 10xcryptodev** [comment]
@UdjinM6 reviewed the points about clear selected outputs and warning message

**Comment by UdjinM6** [comment]
Hmm... `clear()` is pretty disruptive (imagine filling 10 recipients and then switching tabs accidentally). I was thinking about smth like 803f029a8c (and maybe smth like a00c2e98 on top cause silently dropping coins from selection feels kind of weird).

**Comment by 10xcryptodev** [comment]
@UdjinM6 thanks for the contribution, tested and agreed with the better UX. Added the suggested commits

**Comment by PastaPastaPasta** [comment]
test failure looks unrelated, can you force-push and hopefully CI will be happy

**Comment by 10xcryptodev** [comment]
@PastaPastaPasta passed CI check now

**Comment by UdjinM6** [comment]
Found another issue - when running with `--enableprivatesend=0` the PS tab shows "Balance: 0" yet it allows to send PS txes. I think it makes sense to disable the tab (and the corresponding icon/tray menu item) in this case i.e. ef3f107344.

**Comment by 10xcryptodev** [comment]
@UdjinM6 good catch! also tested with -litemode and it's ok. Added the suggestion

**Comment by PastaPastaPasta** [comment]
Will review in the next few days

**Comment by PastaPastaPasta** [comment]
Started a review, when clicking on "PrivateSend" it doesn't become bold and unbold the prior.
dash-core-pr_3442

**Comment by PastaPastaPasta** [comment]
Also, "Use available balance" button doesn't work properly

**Comment by 10xcryptodev** [comment]
@PastaPastaPasta fixed both bugs

**Comment by PastaPastaPasta** [comment]
The 13 test failures we are seeing in the pull request is very likely due to the fact that this PR is based on  de931a25a3fa7ac76aaad17de4517224dbf90de7 (#3433), between #3433 and current develop #3438 #3456. A rebase on latest develop would likely fix this instability, as we aren't seeing these test failures in other PRs.

**Comment by 10xcryptodev** [comment]
@PastaPastaPasta rebase done :+1:

**Comment by PastaPastaPasta** [comment]
Man you must be really unlucky, cause there are still test failures...

**Comment by PastaPastaPasta** [comment]
test passed this time :+1:

**Comment by 10xcryptodev** [comment]
Lucky now :laughing:

**Pull request merged**